### PR TITLE
(WIP) V3

### PR DESCRIPTION
**Pull request by ravanscafi**
## To-do
- [X] PHP 7.1 
- [X] Code Standards
- [X] Fix directories
- [x] Drop connection `pool` (see https://github.com/mongodb/mongo-php-driver/issues/688#issuecomment-349193875)
- [x] Better `update` method (including #122)
- [x] Allow modification of array values directly
- [x] Stop saving fields with value `null`
- [x] Replace `attributes` by other field
- [x] Avoid `first` queries with `null`
- [X] Better relations support (#136)
- [X] Relationship cache on model
- [x] `created_at` and `updated_at` management on embedded relations
- [X] Add to [PHP libraries](https://docs.mongodb.com/ecosystem/drivers/php-libraries/) (see https://github.com/mongodb/docs-ecosystem/pull/463)
- [x] Suggest dependent packages
- [x] Review `Cursor` and `CacheableCursor` (#115)
    - [x] Drop CacheableCursor
- [x] Review `DataMapper` as it does not follow Data Mapper pattern correctly (perhaps drop it?)
- [X] Leverage [`Serializable`](https://secure.php.net/manual/en/class.mongodb-bson-serializable.php)/[`Unserializable`](https://secure.php.net/manual/en/class.mongodb-bson-unserializable.php) classes on saving/retrieving objects implementing [Persistable](http://php.net/manual/pt_BR/class.mongodb-bson-persistable.php).
- [x] Create `ModelInterface`
- [x] Update Readme
- [x] New documentation

## Dependent Packages
- [ ] [Mongolid Laravel](https://github.com/leroy-merlin-br/mongolid-laravel) (see https://github.com/leroy-merlin-br/mongolid-laravel/pull/126)

## Maybe (not breaking changes, so that can be implemented later)
- [ ] Hidden attributes when converting to json #25
- [ ] `refresh()` (#40)
- [ ] `complete()`/`incomplete()` and saving exceptions (for models retrieved with projections)
- [ ] Better relationship support (some reverse relationships that are missing, using laravel collections,  methods like `has`, `is`, etc.)

### CODE REVIEW

File: composer.json [hunk 1]
```diff
@@ -11,31 +11,45 @@
...
     "require-dev": {
-        "mockery/mockery": "^1.0",
-        "satooshi/php-coveralls": "^1.0",
-        "phpunit/phpunit": "^6.4",
-        "squizlabs/php_codesniffer": "^2.0",
-        "sami/sami": "^4.0"
+        "leroy-merlin-br/coding-standard": "^0.1",
+        "mockery/mockery": "^1.2",
+        "php-coveralls/php-coveralls": "^2.1",
+        "phpunit/phpunit": "^6.5",
```
Comment (carusogabriel) [inline composer.json]: We can use PHPUnit 7 here, as we are now minimum requiring PHP 7.1.

**Comment by carusogabriel** [comment]
@ravanscafi What about implementing a `last()` method? The opposite for `first()`.

File: src/Model/HasAttributesTrait.php [hunk 2]
```diff
@@ -0,0 +1,238 @@
...
+        }
+    }
+
+    /**
+     * {@inheritdoc}
+     */
+    public function syncOriginalDocumentAttributes()
+    {
+        try {
+            $this->originalAttributes = unserialize(serialize($this->getDocumentAttributes()));
```
Comment (ravanscafi) [inline src/Model/HasAttributesTrait.php]: note to self: this takes way too many memory, think about alternatives or drop it altogether.

File: README.md [hunk 3]
```diff
@@ -1,85 +1,45 @@
...
-```
-$ composer require leroy-merlin-br/mongolid
-```
+**Note:** If you are working with Laravel, take a look at [mongolid-laravel repository](https://github.com/leroy-merlin-br/mongolid-laravel).
 
-### Requirements
-
-- PHP**7**
+## Requirements
+- PHP **7.1** or superior
```
Comment (carusogabriel) [inline README.md]:
```suggestion
- PHP **7.1** or higher
```

File: composer.lock [hunk 4]
```diff
@@ -0,0 +1,3401 @@
+{
```
Comment (carusogabriel) [inline composer.lock]: I don't see the value of adding the `composer.lock` file to a package, as Composer ignores it by default.

**Pull request closed without merging**